### PR TITLE
Add sat recursive endpoints with index and pagination

### PR DESCRIPTION
**Pull request by raphjaph**
This PR takes the discussion from #2452 and implements the following recursive endpoints:

`/r/sat/:sat` ->  get the first 100 inscription ids on a sat (ascending reinscription order)
`/r/sat/:sat/:page` -> get 100 inscription ids on sat paginated (ascending reinscription)

These two endpoints return:

```json
{
  "ids": [ 
    "c3a05e88abd51708954ce562389478028567184e32fb96ab8d07b672989155b8i0", 
    "c3a05e88abd51708954ce562389478028567184e32fb96ab8d07b672989155b8i1"
  ],
  "page": 2,
  "more": false,
}
```

`/r/sat/:sat/at/:index` -> get a single inscriptions id on a sat by index (-1 is latest, 0 is first inscription)

This returns: 

```json
{  
  "id": "c3a05e88abd51708954ce562389478028567184e32fb96ab8d07b672989155b8i0", 
}
```

`:sat` can be any represenation of a sat including number, degree or letter notation.

### CODE REVIEW

File: tests/server.rs [hunk 1]
```diff
@@ -409,3 +409,44 @@ fn all_endpoints_in_recursive_directory_return_json() {
...
+  let rpc_server = test_bitcoincore_rpc::spawn();
+  create_wallet(&rpc_server);
+
+  rpc_server.mine_blocks(2);
+
+  let server = TestServer::spawn_with_args(&rpc_server, &["--index-sats"]);
+
+  assert_eq!(
+    server
+      .request("/r/inscriptions/sat/5000000000")
```
Comment (elocremarc) [inline tests/server.rs:424]:
```diff
-      .request("/r/inscriptions/sat/5000000000")
+      .request("/r/ids/sat/5000000000")
```

File: src/subcommand/server.rs [hunk 2]
```diff
@@ -229,6 +229,11 @@ impl Server {
         .route("/r/blockheight", get(Self::block_height))
         .route("/r/blocktime", get(Self::block_time))
         .route("/r/metadata/:inscription_id", get(Self::metadata))
+        .route("/r/inscriptions/sat/:number", get(Self::sat_inscriptions))
+        .route(
+          "/r/inscriptions/sat/:number/:from",
```
Comment (elocremarc) [inline src/subcommand/server.rs:234]:
```diff
-          "/r/inscriptions/sat/:number/:from",
+          "/r/ids/sat/:number/:from",
```

File: tests/server.rs [hunk 3]
```diff
@@ -409,3 +409,44 @@ fn all_endpoints_in_recursive_directory_return_json() {
...
+  );
+
+  for _ in 0..10 {
+    inscribe(&rpc_server);
+  }
+
+  let server = TestServer::spawn_with_args(&rpc_server, &["--index-sats"]);
+
+  let response = server
+    .request("/r/inscriptions/sat/5000000000")
```
Comment (elocremarc) [inline tests/server.rs:441]:
```diff
-    .request("/r/inscriptions/sat/5000000000")
+    .request("/r/ids/sat/5000000000")
```

File: src/subcommand/server.rs [hunk 4]
```diff
@@ -229,6 +229,11 @@ impl Server {
         .route("/r/blockheight", get(Self::block_height))
         .route("/r/blocktime", get(Self::block_time))
         .route("/r/metadata/:inscription_id", get(Self::metadata))
+        .route("/r/inscriptions/sat/:number", get(Self::sat_inscriptions))
```
Comment (elocremarc) [inline src/subcommand/server.rs:232]:
```diff
-        .route("/r/inscriptions/sat/:number", get(Self::sat_inscriptions))
+        .route("/r/inscriptions/sat/:number", get(Self::sat_inscriptions))
```

**Review by elocremarc — COMMENTED** [review]
Should shorten route name to save bytes.

File: src/subcommand/server.rs [hunk 5]
```diff
@@ -1441,6 +1451,47 @@ impl Server {
...
+  async fn sat_inscriptions(
+    Extension(index): Extension<Arc<Index>>,
+    Path(sat): Path<u64>,
+  ) -> ServerResult<Json<SatInscriptionsJson>> {
+    Self::sat_inscriptions_paginated(Extension(index), Path((sat, 0))).await
+  }
+
+  async fn sat_inscriptions_paginated(
+    Extension(index): Extension<Arc<Index>>,
+    Path((sat, page_index)): Path<(u64, u64)>,
```
Comment (casey) [inline src/subcommand/server.rs:1463]:
```suggestion
    Path((sat, page)): Path<(u64, u64)>,
```

File: src/subcommand/server.rs [hunk 6]
```diff
@@ -1441,6 +1451,47 @@ impl Server {
...
+        "this server has no sat index".to_string(),
+      ));
+    }
+
+    let (ids, more) = index.get_inscription_ids_by_sat_paginated(Sat(sat), 100, page_index)?;
+
+    Ok(Json(SatInscriptionsJson {
+      ids,
+      more,
+      page: page_index,
```
Comment (casey) [inline src/subcommand/server.rs:1476]:
```suggestion
      page,
```

File: src/subcommand/server.rs [hunk 7]
```diff
@@ -1441,6 +1451,47 @@ impl Server {
...
+    Extension(index): Extension<Arc<Index>>,
+    Path((sat, page_index)): Path<(u64, u64)>,
+  ) -> ServerResult<Json<SatInscriptionsJson>> {
+    if !index.has_sat_index() {
+      return Err(ServerError::NotFound(
+        "this server has no sat index".to_string(),
+      ));
+    }
+
+    let (ids, more) = index.get_inscription_ids_by_sat_paginated(Sat(sat), 100, page_index)?;
```
Comment (casey) [inline src/subcommand/server.rs:1471]:
```suggestion
    let (ids, more) = index.get_inscription_ids_by_sat_paginated(Sat(sat), 100)?;
```

File: src/subcommand/server.rs [hunk 8]
```diff
@@ -1441,6 +1451,47 @@ impl Server {
     })
   }
 
+  async fn sat_inscriptions(
+    Extension(index): Extension<Arc<Index>>,
+    Path(sat): Path<u64>,
```
Comment (casey) [inline src/subcommand/server.rs:1456]:
```suggestion
    Path(sat): Path<Sat>,
```

File: src/subcommand/server.rs (hunk 5 above)
Comment (casey) [inline src/subcommand/server.rs:1463]:
```suggestion
    Path((sat, page_index)): Path<(Sat, u64)>,
```

File: src/subcommand/server.rs [hunk 9]
```diff
@@ -1441,6 +1451,47 @@ impl Server {
...
+    Ok(Json(SatInscriptionsJson {
+      ids,
+      more,
+      page: page_index,
+    }))
+  }
+
+  async fn sat_inscription_at_index(
+    Extension(index): Extension<Arc<Index>>,
+    Path((sat, inscription_index)): Path<(u64, isize)>,
```
Comment (casey) [inline src/subcommand/server.rs:1482]:
```suggestion
    Path((sat, inscription_index)): Path<(Sat, isize)>,
```

File: tests/server.rs [hunk 10]
```diff
@@ -409,3 +409,114 @@ fn all_endpoints_in_recursive_directory_return_json() {
 
   assert!(server.request("/blockhash/2").json::<String>().is_err());
 }
+
+#[test]
+fn sat_recursive_endpoints() {
```
Comment (casey) [inline tests/server.rs:414]: Let's make this an `#[ignore]` test.

**Review by casey — CHANGES_REQUESTED** [review]

File: src/index.rs [hunk 11]
```diff
@@ -1053,20 +1082,55 @@ impl Index {
...
+      rtx.open_table(SEQUENCE_NUMBER_TO_INSCRIPTION_ENTRY)?;
 
-    ids.sort_by_key(|(sequence_number, _id)| *sequence_number);
+    let sat_to_sequence_number = rtx.open_multimap_table(SAT_TO_SEQUENCE_NUMBER)?;
 
-    Ok(ids.into_iter().map(|(_sequence_number, id)| id).collect())
+    if inscription_index < 0 {
+      sat_to_sequence_number
+        .get(&sat.n())?
+        .nth_back(inscription_index.abs_diff(0))
```
Comment (casey) [inline src/index.rs:1117]: This should be tested, and I think you'll need to do `(inscription_index - 1).abs_diff(0)`

**Review by casey — CHANGES_REQUESTED** [review]

File: src/subcommand/server.rs (hunk 8 above)
Comment (elocremarc) [reply src/subcommand/server.rs:1456]: Do we want Sat? Raph had me take it out of mine. Would also require more complexity for any other alternative clients supporting this endpoint.

Comment (elocremarc) [reply src/subcommand/server.rs:1456]: wrong function disregard.

File: src/index.rs (hunk 11 above)
Comment (raphjaph) [reply src/index.rs:1117]: should be +1

**Comment by elocremarc** [comment]
Why are we doing `/r/sat/:sat/at/:index`
`at` is to many bytes its not needed and why are you disregarding my PR? This pretty much kills my entire motivation to help at all on this repo kinda a little frustrated right now to be honest because I worked hard on that PR and you just go do your own.

**Comment by casey** [comment]
> Why are we doing `/r/sat/:sat/at/:index` `at` is to many bytes its not needed

We have three routes: Get inscriptions on a sat, get inscriptions on a sat with page index, and get a single inscription on a sat by index.

We were discussing names today, and the three names that we thought were best are:

```
/sat/:sat # get first page of inscriptions on a sat
/sat/:sat/:page # get `page` of inscriptions on a sat
/sat/:sat/at/:index # get a single sat at an index
```

We thought `/sat/:sat/at/:index` was a less ambiguous option.

> and why are you disregarding my PR? This pretty much kills my entire motivation to help at all on this repo kinda a little frustrated right now to be honest because I worked hard on that PR and you just go do your own.

I understand your frustration, but opening a PR is not a guarantee that it will be merged, and not a guarantee that someone else won't open and merge a PR for the same functionality. Sometimes PRs would be hard to get into mergable shape, so we will write our own, when we want to get the feature out quickly.

Raph and I are super rate limited, we have an infinite amount of stuff to do, and very limited bandwidth to do it. I'd like to do better with this in the future. I don't know what the best way to handle this is, in general. We want to welcome contributions and contributors, but we have to prioritize getting `ord` into the best shape possible.

**Comment by elocremarc** [comment]
> We thought /sat/:sat/at/:index was a less ambiguous option.

Isn't that what documentation is for it uses zero blockspace? We had a lot of discursion of reducing the byte limit the sat number already takes up quite a bit of characters. 

> Sometimes PRs would be hard to get into mergable shape, so we will write our own, when we want to get the feature out quickly.

Understood glad I was able to help move this discussion along opening a PR.

> We want to welcome contributions and contributors, but we have to prioritize getting ord into the best shape possible.

How do we get people up to speed so their PRs can be closer to being mergable? Some of it might take a bit of hand holding and a bit of bandwidth however the more people we get familiar with the codebase the better this will become. Streams are a big help.

**Comment by elocremarc** [comment]
```rust
        .route("/r/sat/:sat", get(Self::sat_inscriptions))
        .route("/r/sat/:sat/:page", get(Self::sat_inscriptions_paginated))
        .route("/r/sat/:sat/at/:index", get(Self::sat_inscription_at_index))
```

I see now that `/r/sat/:index` would clash with the pagination route without the at.  Is the first one needed? Can we not bring back the sat info to get the output of the sat with that one. Because you can get at the 100 with the page route however this kind of closes off having sat info in the future if we ever want to add that. I think knowing the output is very desirable and non deterministic.

**Comment by casey** [comment]
We were thinking of adding something like `/r/location/:sat` to get the current location of a sat.

**Comment by elocremarc** [comment]
> We were thinking of adding something like /r/location/:sat to get the current location of a sat.

`/r/info/:sat` Do we not want to add rarity or date? Can you find rarity with a JS library? Date could be cool to make a gallery of older sat inscriptions. Info is also shorter. I know we don't want to make an API however for web scrapers so I see the balance with that decision.

**Comment by casey** [comment]
Rarity and date can both be calculated with a JS library, so I don't think we'd add them.

**Comment by elocremarc** [comment]
> We were thinking of adding something like /r/location/:sat to get the current location of a sat.

LGTM then!

**Review by casey — APPROVED** [review]
LGTM!

**Pull request merged**